proof/access-control/RISCV64/ArchArch_AC.thy

```diff
@@ -0,0 +1,1995 @@
+             pas_cap_cur_auth aag (ArchObjectCap (ASIDPoolCap frame base)) \<and>
+             (\<forall>x. asid_high_bits_of x = asid_high_bits_of base \<longrightarrow> is_subject_asid aag x)"
+             in hoare_strengthen_post)
+    apply (simp add: page_bits_def)
+    apply (wp add: delete_objects_pspace_no_overlap hoare_vcg_ex_lift
+                   delete_objects_descendants_range_in delete_objects_invs_ex
+                   delete_objects_pas_refined
+              del: Untyped_AI.delete_objects_pspace_no_overlap
+           | simp add: page_bits_def)+
+   apply clarsimp
+   apply (rename_tac s idx)
+   apply (frule untyped_cap_aligned, simp add: invs_valid_objs)
+   apply (clarsimp simp: cte_wp_at_def aag_cap_auth_def ptr_range_def pas_refined_refl
+                         cap_links_asid_slot_def cap_links_irq_def obj_bits_api_def
+                         default_arch_object_def retype_addrs_def conj_ac
+                         invs_psp_aligned invs_valid_pspace invs_vspace_objs invs_arch_state)
+   apply (rule conjI, force intro: descendants_range_caps_no_overlapI simp: cte_wp_at_def)
+   apply (rule conjI, clarsimp simp: max_free_index_def)
+   apply (prop_tac "valid_cap (UntypedCap False frame pageBits idx) s")
+    apply (clarsimp simp: get_cap_caps_of_state)
+    apply (simp add: Untyped_AI.caps_of_state_valid)
+   apply (clarsimp simp: free_index_of_def max_free_index_def valid_cap_def)
+   apply (rule conjI)
+    apply (cut_tac s=s and ptr="(parent_ptr, parent_idx)" in cap_refs_in_kernel_windowD)
+      apply ((fastforce simp: caps_of_state_def cap_range_def)+)[3]
+   apply (fastforce simp: x_power_minus_1 is_aligned_no_overflow')
+  apply (clarsimp simp: valid_aci_def authorised_asid_control_inv_def cte_wp_at_caps_of_state)
+  apply (rule conjI)
+   apply (drule untyped_slots_not_in_untyped_range)
+        apply (erule empty_descendants_range_in)
+       apply (simp add: cte_wp_at_caps_of_state)
+      apply simp
+     apply simp
+    apply (rule subset_refl)
+   apply (simp add: page_bits_def)
+  apply (frule_tac x=x in bspec)
+   apply (simp add: is_aligned_no_overflow)
+  apply (clarsimp simp: ptr_range_def invs_psp_aligned invs_valid_objs aag_cap_auth_def
+                        descendants_range_def2 empty_descendants_range_in page_bits_def
+                        pas_refined_refl cap_links_asid_slot_def label_owns_asid_slot_def
+                        cap_links_irq_def range_cover_def obj_bits_api_def pageBits_def
+                        default_arch_object_def and_mask_eq_iff_shiftr_0 mask_zero)
+  apply (subst is_aligned_neg_mask_eq[THEN sym], assumption)
+  apply (intro conjI; fastforce intro: empty_descendants_range_in)
+  done
+
+(* FIXME AC: more general than pt_slot_offset_id, replace if possible *)
+lemma pt_slot_offset_id':
+  "\<lbrakk> is_aligned pt_ptr pt_bits; mask pt_bits && (xa << pte_bits) = xa << pte_bits \<rbrakk>
```

Comment:
Please rename `xa`. If `pt_slot_offset_id` can actually be replace, that would be cool. If not it can probably be derived from this one, at least.